File: 2021/07/23/remove_duplicates.py
"""
https://leetcode.com/problems/remove-duplicates-from-sorted-list/
Author: Andrew Jarombek
Date: 7/23/2021
"""

from typing import Optional


class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


def delete_duplicates(head: ListNode) -> Optional[ListNode]:
    if head is None:
        return None

    previous = head
    current = head.next

    while current is not None:
        if previous.val == current.val:
            current = current.next
            previous.next = current
        else:
            previous = current
            current = current.next

    return head


if __name__ == '__main__':
    result_list = delete_duplicates(head=ListNode(1, ListNode(1, ListNode(2, ListNode(3, ListNode(3))))))
    assert result_list.val == 1
    assert result_list.next.val == 2
    assert result_list.next.next.val == 3
    assert result_list.next.next.next is None

File: 2022/02/08/rust/src/remove_duplicates.rs
// https://leetcode.com/problems/remove-duplicates-from-sorted-list/
// Author: Andrew Jarombek
// Date: 2/8/2022

#[derive(PartialEq, Eq, Clone, Debug)]
pub struct ListNode {
    pub val: i32,
    pub next: Option<Box<ListNode>>
}

impl ListNode {
    #[inline]
    fn new(val: i32) -> Self {
        ListNode {
            next: None,
            val
        }
    }
}

pub fn delete_duplicates(head: Option<Box<ListNode>>) -> Option<Box<ListNode>> {
    if head.is_none() {
        return head;
    }

    let mut n = head;
    let mut node = n.as_mut().unwrap();

    while let Some(next) = node.next.as_mut() {
        if node.val == next.val {
            node.next = next.next.take();
        } else {
            node = node.next.as_mut().unwrap();
        }
    }

    return n;
}